data_warehouse/services/coverage_statistics_service.py

```diff
@@ -0,0 +1,202 @@
+'''培训覆盖率服务'''
+from django.utils.timezone import now
+from django.db.models import Count
+from training_record.models import Record
+from training_program.models import Program
+from auth.models import User, Department
+from auth.services import DepartmentService
+from infra.exceptions import BadRequest
+
+
+class CoverageStatisticsService:
+    '''培训覆盖率统计服务'''
+    TITLES = (
+        '教授', '副教授', '讲师', '助教', '教授级工程师',
+        '高级工程师', '工程师', '研究员', '副研究员', '助理研究员')
+    AGES = (
+        ('35岁及以下', [0, 35]), ('36-45岁', [36, 45]),
+        ('46-55岁', [46, 55]), ('56岁及以上', [56, 150]))
+
+    @staticmethod
+    def teacher_coverage():
+        '''专任教师培训覆盖率统计'''
+
+    @staticmethod
+    def groupby_ages(users_qs):
+        '''按照年龄段分组
+
+        Parameters
+        -------
+        users_qs: QuerySet
+            待分组的User查询集
+
+        Returns
+        ------
+        list of dict
+            按年龄分组后的结果，list包含的元素dict定义为： {
+                age_range: string,
+                    年龄段，例如 35岁及以下。
+                coverage_count: int,
+                    参数users_qs包含的用户中，该年龄段对应的用户人数。
+                total_count: int
+                    系统里该年龄段对应的用户总人数。
+            }
+        '''
+        data = []
+        for age_range in CoverageStatisticsService.AGES:
+            label = age_range[0]
+            data.append(
+                {
+                    'age_range': label,
+                    'coverage_count': users_qs.filter(
+                        age__range=age_range[1]).count(),
+                    'total_count': User.objects.filter(
+                        age__range=age_range[1]).count()
+                }
+            )
+        return data
+
+    @staticmethod
+    def groupby_titles(users_qs):
+        '''按照职称分组
+
+        Parameters
+        -------
+        users_qs: QuerySet
+            待分组的User查询集
+
+        Returns
+        ------
+        list of dict
+            按职称分组后的结果，list包含的元素dict定义为： {
+                title: string,
+                coverage_count: int,
+                    参数users_qs包含的用户中，该职称对应的用户人数。
+                total_count: int
+                    系统里该职称对应的用户总人数。
+            }
+        '''
+        data = []
+        coverage_titles = users_qs.values('technical_title').annotate(
+            coverage_count=Count('technical_title'))
```

Comment:
IRRC, there are plenty of `technical_title`s, and we should categorize titles we don't care into `others`, right?

Reply:
> IRRC, there are plenty of `technical_title`s, and we should categorize titles we don't care into `others`, right?

The function in this service is intended to provide raw data, if caller only needs subset of all technical titles, I think caller or some view related functions should be responsible for doing these jobs you mentioned, because it' s caller but current function that knows what  matter.

Reply:
I had enhanced  the function, which is able to support optional `interesting group lists` now.